### PR TITLE
Layout: Add default fallback gap value in block.json, and set to 2em for Columns blocks

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

This is an exploration into a potential fix for the gap regression in https://github.com/WordPress/gutenberg/issues/40952 (it doesn't address the tablet breakpoint change raised in that issue).

In this PR we look at:

* Adding a setting in `block.json` where blocks can declaratively provide a default fallback value for properties of the spacing block support. This PR only implements support for the `blockGap` property.
* In the Layout block support, look up the block's definition, and see if it provides its own fallback value for blockGap, and use it if available.
* For the Columns block we set the default fallback blockGap value to `2em` to be consistent with backwards compatibility.
* If no default value is provided we use the fallback value of `0.5em`.

The idea here is that this new fallback value isn't in `theme.json` because it's a part of the _block_'s styles — not the theme's, and is intended to form part of the base styling of the block, upon which themes can make their own changes.

Kudos @ramonjd for collaborating on the idea in this PR!

## Why?

In #40952 it was flagged that the Columns block used to have a default gap value of `2em`. This PR explores seeing whether we can declaratively set the fallback value for blocks in their `block.json` file — outside of the concept of theming. The gap of `2em` was originally removed in https://github.com/WordPress/gutenberg/pull/37436 when the Columns block was updated to opt-in to the Layout block support — it's the Layout block support that provides the default value of `0.5em`, so this PR gives the Layout support a little more power to look up and output a fallback value that is suitable for the block being rendered.

## How?

* Add a `default` prop to the features of the `spacing` object in `block.json` to specify a default fallback value. E.g. `spacing.blockGap.default`.
* Add a default value of `2em` to the Columns block, with a default fallback of `0.5em`.
* Use this default value in the Layout block support.

## Testing Instructions

1. Using a classic theme like TwentyTwentyOne, add a columns block to a post or page.
2. Notice that prior to this PR, the gap between columns is `0.5em`. With this PR applied it should be `2em` in both the editor and on the front end of the site.
3. Using a blocks-based theme, check that the block spacing support still functions as expected (E.g. you can set a block spacing value in the post editor when editing a columns block). Note that block-level block spacing in global styles / theme.json is not yet supported, but the issue is raised in https://github.com/WordPress/gutenberg/issues/39789

## Screenshots or screencast 

| Before | After |
| --- | --- |
| image | image |

### CODE REVIEW

File: packages/block-library/src/columns/block.json
Comment (ramonjd) [inline packages/block-library/src/columns/block.json]: I wonder if it makes sense to use the proposed block.json changes from https://github.com/WordPress/gutenberg/pull/34180/files#diff-dba084b81a05295345ce9149af1a17ab073ddef43a393e851bcd3bdae85d3f32R81 for default values. (PR: **Prototype: merge block CSS with theme.json styles** #34180)

I don't know how far that particular investigation will go, but it could save some lines to "join forces"
cc @oandregal

Comment (andrewserong) [reply packages/block-library/src/columns/block.json]: Oh, neat idea!

There's quite a lot of similarity here — for the case of `blockGap` we only want to use the value as a fallback value if none is set anywhere else, rather than always outputting it, so that might be a subtle difference between a default / fallback value and a value set in `blockStyles`. But keen for feedback or ideas on how to combine things if you have any @oandregal!

I suppose the caveat here is that this PR will likely need to be backported to 6.0 RC whereas it looks like #34180 is a bit more forward-looking and something potentially for 6.1?

Comment (ramonjd) [reply packages/block-library/src/columns/block.json]: > I suppose the caveat here is that this PR will likely need to be backported to 6.0 RC whereas it looks like https://github.com/WordPress/gutenberg/pull/34180 is a bit more forward-looking and something potentially for 6.1?

Ah, very good point. 

`__experimentalDefaultValues` does also do what it says on the tin as you say.

**Comment by ramonjd** [comment]
> In https://github.com/WordPress/gutenberg/issues/40952 it was flagged that the Columns block used to have a default gap value of 2em. This PR explores seeing whether we can declaratively set the fallback value for blocks in their block.json file — outside of the concept of theming.

I'm trying think of other use cases for which this change would be useful. Basically, I'm keen on this PR and want to shore up my confirmation bias 🥴 but I can't right now 🤔 

We're trying to set a style fallback value for a block; a value that can be referenced across the application where no other exists. 

Right now, it's the layout implementation that needs to know a particular block's gap value before it arbitrarily sets it own.

The way the styles are output means that we can't solve this using CSS alone, e.g., by adding `gap: 2em` to `.wp-columns` since it will _always_ be overwritten by the layout styles regardless of whether there's a gap value set or not. 

We want layout to overwrite such style _when a gap value has been set by the user_, but not necessarily when **no** gap value has been set by the user.

It's this maze that has brought us here. So discounting a total refactor of _[insert unknown code here]_, what this PR proposes seems to be the most straight forward approach.

The main attraction for me is that we'll have a place to store Gutenberg block style fallbacks outside CSS files, removing the need for complex CSS overrides if we ever need to change other values in the future.

I guess my only hesitation would be to extend the block API with something that is intended to address a special case.


cc @gziolo asking you because I've seen your handle all over the block.json metadata 😄

File: packages/block-library/src/columns/block.json
Comment (youknowriad) [reply packages/block-library/src/columns/block.json]: Why not just `blockGap: { default: "something" }` instead of `blockGap: true`? (We do these object configs for a lot of block supports already)

File: packages/block-editor/src/layouts/flex.js
Comment (youknowriad) [inline packages/block-editor/src/layouts/flex.js:11]: what about flow.js? no fallback support there?

Comment (andrewserong) [reply packages/block-editor/src/layouts/flex.js:11]: I noticed that too when I went to work on this feature — the Flow layout currently doesn't use any fallback value at all, so I didn't inject one where there isn't one already (both in `layout.php` and `flow.js`). From looking over the code again, it appears that blockGap styles in the Flow layout are only ever output if the blockGap has been opted into, in which case there'll always be a value, so a fallback value will never be reached. E.g. in the following block:

https://github.com/WordPress/gutenberg/blob/6de16981c7fef3db7cc4a9d6cfcb0ad85c916c62/packages/block-editor/src/layouts/flow.js#L168

It was the Flex layout that currently has a `0.5em` hard-coded fallback, and where I think this feature is useful based on the current implementations.

But, please do let me know if I missed something here! 🙂

File: packages/block-library/src/columns/block.json
Comment (andrewserong) [reply packages/block-library/src/columns/block.json]: The individual feature properties of the `spacing` support are currently `boolean | array` and the array is treated as a list of sides that are opted into, with logic like the following function (and the below `useIsDimensionsSupportValid`: https://github.com/WordPress/gutenberg/blob/2ce8e8d6ce5572518750c2e10cee9f2aa22d595c/packages/block-editor/src/hooks/dimensions.js#L167

It'd be a more complex change to refactor those props to also support an object (with a nested `sides` key), so I thought using a parallel object to set default values might make for a cleaner change given that this PR could be backported. There's a precedent in the Layout support, too, which has a default object and then within it, it specifies the properties:

https://github.com/WordPress/gutenberg/blob/56152c6037b46ff98de4e0247e5fb9eb91fca85d/packages/block-library/src/columns/block.json#L45

We could always rename this from `__experimentalDefaultValues` to `default`, and it'd look like the following:

```
		"spacing": {
			"blockGap": true,
			"margin": [ "top", "bottom" ],
			"padding": true,
			"__experimentalDefaultControls": {
				"padding": true
			},
			"default": {
				"blockGap": "2em"
			}
		},
```

What do you think?

**Comment by andrewserong** [comment]
Thanks for you thoughts @ramonjd!

> I guess my only hesitation would be to extend the block API with something that is intended to address a special case.

Yes, that's my main hesitation too — whether or not it's useful beyond this special case. I guess that's also one of the reasons I leaned toward listing it as an `__experimental` prop, so that we could keep it in while we need the feature, and as we continue to explore options for refactoring the Layout support (like the presets-based approaches and so on), we could then remove it altogether when it's no longer useful. But, of course, I know it's hard to remove things once we've added them! 😅

File: packages/block-library/src/columns/block.json
Comment (ramonjd) [reply packages/block-library/src/columns/block.json]: Also we'd have to find a model to define default side values too, right? Well, maybe not initially. Still that would involve some data massaging I suppose.

```json
"margin": [ "top", "bottom" ]
```

To something like

```json
"margin": { default: { "top": "1em", ... } }
```

🤷 

> The individual feature properties of the spacing support are currently boolean | array and the array is treated as a list of sides that are opted into

It's possibly not the cleanest way, but I guess we _could_ check for the existence of a `default` key (?) if we had to?

```js
	if (
		! support ||
		typeof support[ feature ] === 'boolean' ||
		!! support?.[ feature ]?.default
	) {
		return;
	}
```

I'm not saying it's pretty, and I kinda also wonder if the complexity is worth it.

**Comment by ramonjd** [comment]
This is working pretty well for me:

**Classic theme**
Screen Shot 2022-05-18 at 9 59 34 am

**Block theme**
Screen Shot 2022-05-18 at 9 57 29 am

> so that we could keep it in while we need the feature, and as we continue to explore options for refactoring the Layout support (like the presets-based approaches and so on),

I can't fault that logic. In some situations we need the freedom to explore ideas, which was part of the point of `__experimental` I believe

**Comment by peterwilsoncc** [comment]
I did some testing manually and this appears to be working nicely.

I know some people have strong feelings about putting `__experimental` items in to Core but it's an ongoing discussion so let's not call it a blocker.

File: packages/block-library/src/columns/block.json
Comment (youknowriad) [reply packages/block-library/src/columns/block.json]: > There's a precedent in the Layout support, too, which has a default object and then within it, it specifies the properties:

Yeah, that's the one I was thinking about, since you can also do. `__experimentalLayout: true` (a defined object is the same thing as "true"). I don't think we should bother changing the others (padding, margin) now, but would be good to decide on an approach. I'd vote for the object approach personally but wouldn't mind the alternative used here.

**Comment by ndiego** [comment]
Are we still trying to get this into 6.0, or are we looking at 6.0.1? This is basically the last item on the [6.0 Project Board](https://github.com/orgs/WordPress/projects/27), but since RC3 is already out, I was planning to move to the 6.0.1 board. However given the impact of the original issue and all the activity here, I'm second guessing myself. 😅 Just let me know either way, thanks!

**Comment by gziolo** [comment]
@peterwilsoncc highlighted this issue as one of the bug fixes we want to get into WP 6.0 RC 4 which will probably happen on Friday. See https://wordpress.slack.com/archives/C035NHLFUSY/p1652848301916789 (link requires registration at https://make.wordpress.org/chat).

**Comment by ndiego** [comment]
Perfect, thanks @gziolo!

**Comment by gziolo** [comment]
@andrewserong, do you plan to apply any feedback from the review or is it good enough as a temporary solution to ship in WordPress 6.0 next week? From my perspective, this PR can target directly `wp/6.0` branch and contain minimal changes that are scoped to the Columns block and have even specific flag in block support like `supports.__experimentalFallbackDefaultGap: "2em"`. This way, you can work on the more general API separately that will get tested in the Gutenberg plugin first before landing in WordPress core.

**Review by ramonjd — APPROVED** [review]
I've tested this in block themes, core classic themes going back to 2011, and a few custom themes e.g., Techine, OceanWP, Neve, Livro, Conceptly

Screen Shot 2022-05-19 at 9 51 36 am

> this PR can target directly wp/6.0 branch and contain minimal changes that are scoped to the Columns block and have even specific flag in block support like supports.__experimentalFallbackDefaultGap: "2em". 

This is interesting! So we'd add these changes for the wp-6.0 branch only?

Sounds like a neat way to give us some breathing room and finesse the API for 6.1. 👍 

I suppose it'd make implementation in this PR simpler if we had a single block.json property `__experimentalFallbackDefaultGap ` for columns block.json only, but would it make much a difference @andrewserong ?

File: packages/block-library/src/columns/block.json
Comment (andrewserong) [reply packages/block-library/src/columns/block.json]: Thanks for the feedback here! It turns out it wasn't as complex to deal with as I'd thought given that we only appear to look up which sides are opted into in the one place (`useCustomSides`).

In https://github.com/WordPress/gutenberg/pull/41098/commits/a2cc15843c315601feabfeb3bee8dedcac917b9f, I've gone with the approach of:

```
		"spacing": {
			"blockGap": {
				"default": "2em"
			},
```

And we can also specify a `sides` key, too, in that object if we need to.

**Comment by andrewserong** [comment]
Thanks for the feedback, folks!

I've updated the PR in https://github.com/WordPress/gutenberg/pull/41098/commits/a2cc15843c315601feabfeb3bee8dedcac917b9f to go with [Riad's suggestion](https://github.com/WordPress/gutenberg/pull/41098#discussion_r874637730) of moving the `default` value to be an object property of `spacing.blockGap`, as this feels more consistent with other approaches (e.g. layout) and seems to neaten things up a little bit.

> From my perspective, this PR can target directly wp/6.0 branch

I _think_ the way it's implemented now, if folks agree, could happily merge into `trunk` and then we could cherry pick it into that branch? Ideally it'd be nice to have the feature work the same in `trunk` and `wp/6.0` so that we don't unexpectedly have different behaviour.

If there isn't consensus on landing the PR in its current shape, though, I'm happy to open up a separate PR targeting `wp/6.0` tomorrow morning, scoping it to just the columns block.

Thanks again! 🙇

File: packages/block-editor/src/hooks/dimensions.js
Comment (ramonjd) [inline packages/block-editor/src/hooks/dimensions.js:176]: Nice. Is this to catch axial values like `margin: (2) ['top', 'bottom']`?

Would it worth writing a small comment to that effect as a hint to others?

File: packages/block-library/src/columns/block.json
Comment (ramonjd) [inline packages/block-library/src/columns/block.json:33]: I like!

**Comment by ramonjd** [comment]
Just noting here that there's a [PR](https://github.com/WordPress/gutenberg/pull/41123) to address the other regression reported in https://github.com/WordPress/gutenberg/issues/40952

It reinstates the "two column" stack for tablet view. It's not as important as this PR I think, that is the visual impact is not as great. Only if it's decided that it is important and needs back-porting (if that's at all possible at this stage).

Thanks!

**Comment by glendaviesnz** [comment]
This tested as expected for me across block and non-block themes, except for one theme that is overriding the left margin of the last child (which is how the column gap used to be set). In this case, with this fix they end up with a much wider gap than before:

margin

**Comment by andrewserong** [comment]
Thanks for testing @glendaviesnz!

> except for one theme that is overriding the left margin of the last child (which is how the column gap used to be set)

I suppose the question here is whether we're confident enough in having the default fallback of `2em` (and fixing the regression) but requiring theme developers who use these overrides to either a) update their CSS to target `gap` on the parent columns block (with the right specificity), or b) add / update their `theme.json` file to set a block `blockGap` spacing appropriate for their use case.

File: packages/block-editor/src/hooks/dimensions.js
Comment (andrewserong) [reply packages/block-editor/src/hooks/dimensions.js:176]: Yep! That's the case: I've added an explanatory comment in https://github.com/WordPress/gutenberg/pull/41098/commits/e779a9ff8f52d550eee6bb2c147f30d10ea0c015 to hopefully make it clearer.

**Comment by andrewserong** [comment]
I'm about to wrap up for the day — I think the PR is in a reasonable shape as a proposal for fixing the regression, with the caveats that Glen mentioned that with the changes in 6.0 (moving to `gap` instead of margins between columns) some themes will need to update their CSS. If there's agreement on the approach, feel free to merge this / cherry pick into 6.0, or let me know if we should take a different approach, and I can follow-up tomorrow.

Thanks, all!

**Comment by ramonjd** [comment]
> I think the PR is in a reasonable shape as a proposal for fixing the regression

Thanks a lot @andrewserong and @glendaviesnz for getting this one ready today! 🙇

File: packages/block-library/src/columns/block.json
Comment (adamziel) [inline packages/block-library/src/columns/block.json]: How about an __experimental property?

```suggestion
				"__experimentalDefault": "2em"
```

Comment (gziolo) [reply packages/block-library/src/columns/block.json]: Good catch @adamziel, It looks like a new API would be a more controversial choice in the context of the WP 6.0 release. Regardless of how we apply that to `wp/6.0` branch, this PR will require documentation changes as a follow-up.

Comment (andrewserong) [reply packages/block-library/src/columns/block.json]: Thanks, yes that's a great idea Adam, I'll update to add an `experimental` prefix 👍

Comment (andrewserong) [reply packages/block-library/src/columns/block.json]: I've added the `__experimental` prefix in: https://github.com/WordPress/gutenberg/pull/41098/commits/7774b7ce4bba13eeace06d8306fcf7a5f7e4539e

If this PR lands in time, @ramonjd and I are happy to write up documentation changes in a follow-up.

**Comment by andrewserong** [comment]
I'm wrapping up for the week now, but feel free to merge this in, anyone, if the approach (and `__experimental` prefix) looks viable. Thanks for all the feedback!

**Review by gziolo — APPROVED** [review]
Let's get this one in for WordPress 6.0 RC 4.

As discussed in the comment, we use `__experimentalDefault` to avoid introducing a new API just before the major WordPress release. However, it feels like it should be perfectly fine to make this API stable in a follow-up PR together with all the necessary documentation changes.

**Comment by adamziel** [comment]
I just cherry-picked this PR to the wp/6.0 branch to get it included in the next release: 497c7ecff6

**Comment by andrewserong** [comment]
Thanks so much @gziolo and @adamziel, much appreciated! 🙇

**Comment by carasmo** [comment]
Please, can someone summarize what I should do to fix this issue? It's horrible to not be able to rely on Gutenberg ... now, if the sites auto update on Flywheel, they will look bonkers.

**Comment by ramonjd** [comment]
> Please, can someone summarize what I should do to fix this issue? It's horrible to not be able to rely on Gutenberg ... now, if the sites auto update on Flywheel, they will look bonkers

Hi @carasmo! 

Do you mean the columns gap? This PR made it into [WordPress 6.0](https://github.com/WordPress/wordpress-develop/pull/2734) so the previous default of `2em` for columns should be there.

**Comment by carasmo** [comment]
Thanks for the reply! I'm using 6.0 and there's no gap between columns.

**Comment by ramonjd** [comment]
> I'm using 6.0 and there's no gap between columns.

Hmm, I just took a look. I was testing WordPress 6.0 running Twenty Twenty-Two theme (latest version of Gutenberg plug **not** activated) and it looks like the new columns gap value is there:

Screen Shot 2022-05-26 at 9 59 31 am

Here's where the value is specified: https://github.com/WordPress/wordpress-develop/blob/6.0/src/wp-includes/blocks/columns/block.json#L32

With the Gutenberg plugin activated however, and running WordPress 6.1-alpha-53448 and Twenty Twenty-Two theme (using this repo's wp-env), the default columns gap value isn't coming through however 🤔

**Comment by ramonjd** [comment]
> With the Gutenberg plugin activated however, and running WordPress 6.1-alpha-53448 and Twenty Twenty-Two theme (using this repo's wp-env), the default columns gap value isn't coming through however 🤔

Withdraw that comment. I hadn't built the plugin properly (`npm run dev` died on me)

So it looks fine to me @carasmo 

Which theme and version of the Gutenberg plugin (if any) are you running? Or is it WordPress 6.0 only?

**Comment by carasmo** [comment]
I'm running a custom theme with 6.0 only.

**Comment by ramonjd** [comment]
> I'm running a custom theme with 6.0 only.

Sorry, I can't reproduce this 🤔 

I'd be great if you could create a new issue with a description of how to reproduce, plus a screenshot of what you're seeing and the name of the theme you're using.

We can then move the discussion to a dedicated place seeing as this PR is merged.

Here's a link to create a new issue: https://github.com/WordPress/gutenberg/issues/new/choose

Thank you!

**Comment by carasmo** [comment]
Note: I don't have a class `.wp-container-...` anywhere. I use a group, then I make columns. The the gap between columns is gone as of 6.0 (figure A). In 5.9.3, there was a margin on the `.wp-block-columns` class (figure B).

Figure A (WP 6):
Screen Shot 2022-05-26 at 5 55 52 AM

Figure B (WP 5.9.3):
Screen Shot 2022-05-26 at 6 08 22 AM

**Comment by pbiron** [comment]
I just ran into this problem as well (while testing several sites before updating to them to 6.0).  These sites use a custom classic (non-block) theme and have a couple of custom dynamic blocks which output markup such as
```
<div class='wp-block-columns custom-class'>
  <div class='wp-block-column another-custom-class'>
     ...
  </div>
  <div class='wp-block-column another-custom-class'>
     ...
  </div>
  <div class='wp-block-column another-custom-class'>
     ...
  </div>
</div>
```
where the custom block relies on the styles that WP < 6.0 included in `wp-includes/css/dist/block-library/style.css`.

My current plan is to update the theme to include
```
@media (min-width: 782px) {
	.wp-block-columns:not( .is-not-stacked-on-mobile ) > .wp-block-column:not( :first-child ) {
		margin-left: var( --wp--style--block-gap, 2em );
	}
}
```
as a short-term workaround so that I can get the sites updated to 6.0.

**Comment by carasmo** [comment]
Yes, I'm going to have to go into many sites and add something in the additional css to address this. In the current site I'm working on, I'm doing the following. Soon I'm going to make my own column block and avoid these time suckers.

```
.theme-body-class .wp-block-columns {
	column-gap: var( --column-gap-svp );
} 

@media ( min-width: $col-collapse-min-width ) {

	.theme-body-class .wp-block-columns {
		column-gap: var( --column-gap-lvp );
	} 
	
}

```

**Comment by mburridge** [comment]
Added the Needs Dev Note label in case this needs a dev note for the 6.1 release.

**Pull request merged**
